### ISSUE TITLE
GCC is able to optimize away a loop

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#62560]
I ran into this will trying to reduce code for #60761. This is non-sensical code but maybe there's a possible underlying missing optimization which would actually make sense.

```
int multiCompare(const char *haystack)
{
    for (;;) {
        if (*haystack == ' ' || *haystack == '\0') {
            break;
        }
        else {
            do {
                ++haystack;

                if (*haystack == ' ' || *haystack == '\0') {
                    return -1;
                }
                if (*haystack == '|') {
                    break;
                }
            } while (true);

            ++haystack;
        }
    }

    return -1;
}
```

https://godbolt.org/z/YMEz46hdE

GCC will optimize the loop away and only generate `return -1` for the function.

**Comment by firewave** [comment]
I am wondering if the fact that a loop can be optimized away could be leveraged for a compiler/static analysis warning.

**Comment by pinskia** [comment]
Note GCC is only able to optimize this loop away for C++ code due to it is not valid to optimize it away for C.
